man/man1/hilm.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "HILM 1"
.TH HILM 1 "2015-03-10" "SHTOOLS 3.0" "SHTOOLS 3.0"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "Hilm"
.IX Header "Hilm"
.IP "Hilm \-" 8
.IX Item "Hilm -"
Calculate iteratively the relief along an interface of constant density contrast that corresponds to a given Bouguer anomaly using the algorithm of Wieczorek and Phillips (1998).
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 Hilm (" 18
.IX Item "SUBROUTINE Hilm ("
\&\s-1CILM, BA, GRID, LMAX, NMAX, MASS, R0, RHO, GRIDTYPE, W, PLX, ZERO, FILTER_TYPE, FILTER_DEG, LMAX_CALC \s0)
.RS 4
.IP "REAL*8" 19
.IX Item "REAL*8"
\&\s-1CILM\s0(2, \s-1LMAX+1, LMAX+1\s0), \s-1BA\s0(2, \s-1LMAX+1, LMAX+1\s0), \s-1GRID\s0(\s-1NLAT, NLONG\s0), \s-1MASS, R0, RHO\s0
.IP "REAL*8, \s-1OPTIONAL\s0" 19
.IX Item "REAL*8, OPTIONAL"
\&\s-1PLX\s0(\s-1LMAX+1,\s0\ (\s-1LMAX+1\s0)*(\s-1LMAX+2\s0)/2), \s-1ZERO\s0(\s-1LMAX+1\s0), W(\s-1LMAX+1\s0)
.IP "\s-1INTEGER\s0" 19
.IX Item "INTEGER"
\&\s-1LMAX, NMAX, GRIDTYPE\s0
.IP "\s-1INTEGER, OPTIONAL\s0" 19
.IX Item "INTEGER, OPTIONAL"
\&\s-1FILTER_TYPE, FILTER_DEG, LMAX_CALC\s0
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIHilm\fR is used to solve iteratively for the relief along an interface that corresponds to a given Bouguer anomaly. This is equation 18 of Wieczorek and Phillips (1998) which implicitly takes into consideration the finite-amplitude correction. Each iteration takes as input a guess for the relief (specified by \s-1GRID\s0) and outputs the iteratively improved spherical harmonic coefficients of this relief. These coefficients can then be re-expanded on grid and re-input into this routine as the next guess. For the initial guess, it is often sufficient to use the relief predicted using the first-order \*(L"mass sheet\*(R" approximation, or perhaps zero. The input relief \s-1GRID\s0 can be of one of three type specified by \s-1GRIDTYPE: 1\s0 for Gauss-Legendre grids, 2 for NxN Driscoll-Healy grids, and 3 for Nx2N Driscoll-Healy grids.
.PP
If the algorithm does not converge, one might want to try damping the initial estimate. Alternatively, iterations of the following form have proven successfulin in damping oscilations between successive iterations:
.PP
h3 = (h2+h1)/2
h4 = f(h3)
.PP
It is important to understand that as an intermediate step, this routine calculates the spherical harmonic coefficients of the relief raised to the nth power, i.e., GRID**n. As such, if the input function is bandlimited to degree L, the resulting function will thus be bandlimited to degree L*NMAX. This subroutine implicitly assumes that \s-1LMAX\s0 is greater than or equal to L*NMAX. If this is not the case, then aliasing will occur. In practice, for accurate results, it is found that \s-1LMAX\s0 needs only to be about twice the size of L, though this should be verified for each application. Thus, if the input function is considered to be bandlimited to degree L, the function should be evaluated on a Gauss-Legendre grid corresponding to a maximum degree of about 2*L.
.PP
If the input grid is evaluated on the Gauss-Legendre points, it is necessary to specify the optional parameters W and \s-1ZERO,\s0 or W and \s-1PLX,\s0 which are calculated by a call to \fIPreCompute\fR. If memory is not an issue, the algorithm can be speeded up considerably by inputing the optional array \s-1PLX\s0 of precomputed associated Legendre functions on the Gauss-Legendre nodes. If \s-1PLX\s0 is not specified, then it is necessary to input the optional array \s-1ZERO\s0 that contains the latitudinal Gauss-Legendre quadrature nodes.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "\s-1CILM \s0" 13
.IX Item "CILM "
(output) REAL*8, \s-1DIMENSION \s0(2, \s-1LMAX+1, LMAX+1\s0) \fIor\fR \s-1DIMENSION \s0(2, \s-1LMAX_CALC+1, LMAX_CALC+1\s0)
.Sp
An estimate of the real spherical harmonic coefficients (geodesy normalized) of relief along an interface with density contrast \s-1RHO\s0 that satisfies the Bouguer anomaly \s-1BA.\s0 The degree zero term corresponds to the mean radius of the relief.
.IP "\s-1BA\s0" 13
.IX Item "BA"
(input) REAL*8, \s-1DIMENSION \s0(2, \s-1LMAX+1, LMAX+1\s0) \fIor\fR \s-1DIMENSION \s0(2, \s-1LMAX_CALC+1, LMAX_CALC+1\s0)
.Sp
The real spherical harmonic coefficients of the Bouguer anomaly referenced to a spherical interface R0.
.IP "\s-1GRID\s0" 13
.IX Item "GRID"
(input) REAL*8, \s-1DIMENSION \s0(\s-1LMAX+1,\s0 2*LMAX+1) for \s-1GRIDTYPE\s0 = 1, \s-1DIMENSION \s0(2*LMAX+2, 2*LMAX+2) for \s-1GRIDTYPE\s0 = 2, \s-1DIMENSION \s0(2*LMAX+2, 4*LMAX+4) for \s-1GRIDTYPE\s0 = 3)
.Sp
The initial estimate for the radii of the interface evaluated a grid corresponding to a function of maximum spherical harmonic degree \s-1LMAX.\s0 This is calculated by a call to either \fIMakeGridGLQ\fR or \fIMakeGridDH\fR This grid must contain the degree\-0 average radius of the interface.
.IP "\s-1LMAX\s0" 13
.IX Item "LMAX"
(input) \s-1INTEGER\s0
.Sp
The spherical harmonic bandwidth of the input relief \s-1GRID.\s0 If \s-1LMAX_CALC\s0 is not set, this determines also the maximum spherical harmonic degree of the output spherical harmonic coefficients of the relief and the input spherical harmonics of the Bouguer anomaly. As a general rule, \s-1LMAX\s0 should be about twice the real spherical harmonic bandwidth of the input function to avoid aliasing.
.IP "\s-1NMAX\s0" 13
.IX Item "NMAX"
(input) \s-1INTEGER\s0
.Sp
The maximum order used in the Taylor-series expansion used in calculating the potential coefficients. As a rule, this should be about 4.
.IP "\s-1MASS\s0" 13
.IX Item "MASS"
(input) REAL*8
.Sp
The mass of the planet in kg.
.IP "R0" 13
.IX Item "R0"
(input) REAL*8
.Sp
The reference radius of the Bouguer anomaly \s-1BA.\s0
.IP "\s-1RHO\s0" 13
.IX Item "RHO"
(input) REAL*8
.Sp
The density contrast of the relief in kg/m^3.
.IP "\s-1GRIDTYPE\s0" 13
.IX Item "GRIDTYPE"
(input) \s-1INTEGER\s0
.Sp
1 = Gauss-Legendre grids, calculated using \fIPreCompute\fR and \fIMakeGridGLQ\fR. 2 = Equally sampled Driscoll-Healy grids, NxN, calculated using \fIMakeGridDH\fR. 3 = Equally spaced Driscoll-Healy grids, Nx2N, calculated using \fIMakeGridDH\fR.
.IP "W" 13
.IX Item "W"
(input) REAL*8, \s-1OPTIONAL DIMENSION \s0(\s-1LMAX+1\s0)
.Sp
The weights used in the Gauss-Legendre quadrature. These are calculated from a call to \fIPreCompute\fR (or alternatively, \fIPreGLQ\fR). If present, one of \s-1PLX\s0 or \s-1ZERO\s0 must also be present.
.IP "\s-1PLX\s0" 13
.IX Item "PLX"
(input) REAL*8, \s-1OPTIONAL, DIMENSION \s0(\s-1LMAX+1, \s0(\s-1LMAX+1\s0)*(\s-1LMAX+2\s0)/2)
.Sp
An array of the associated Legendre functions calculated at the nodes used in the Gauss-Legendre quadrature. These are determined from a call to \fIPreCompute\fR.
.IP "\s-1ZERO\s0" 13
.IX Item "ZERO"
(input) REAL*8, \s-1OPTIONAL, DIMENSION \s0(\s-1LMAX+1\s0)
.Sp
The nodes used in the Gauss-Legendre quadrature over latitude, calculated by a call to \fIPreCompute\fR.
.IP "\s-1FILTER_TYPE\s0" 13
.IX Item "FILTER_TYPE"
(input) \s-1INTEGER, OPTIONAL\s0
.Sp
Apply a filter when calculating the relief in order to minimize the destabilizing effects of downward continuation which amplify uncertainties in the Bouguer anomaly. If 0, no filtering is applied. If 1, use the minimum amplitude filter of Wieczorek and Phillips (1998; equation 19). If 2, use a minimum curvature filter.
.IP "\s-1FILTER_DEG\s0" 13
.IX Item "FILTER_DEG"
(input) \s-1INTEGER, OPTIONAL\s0
.Sp
The spherical harmonic degree for which the filter is 0.5.
.IP "\s-1LMAX_CALC\s0" 13
.IX Item "LMAX_CALC"
(input) \s-1INTEGER, OPTIONAL\s0
.Sp
The maximum degree that will be calculated in the spherical harmonic expansions.
.SH "NOTES"
.IX Header "NOTES"
This routine uses geodesy 4\-pi normalized spherical harmonics that exclude the Condon-Shortley phase; This can not be modified.
.PP
This routine requires the fast Fourier transform library \fI\s-1FFTW\s0\fR, which is available at <http://www.fftw.org>.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIhilmrhoh\fR\|(1), \fIshexpandglq\fR\|(1), \fImakegridglq\fR\|(1), \fIprecompute\fR\|(1), \fIpreglq\fR\|(1), \fIshexpanddh\fR\|(1), \fImakegriddh\fR\|(1), \fIglqgridcoord\fR\|(1)
.PP
<http://shtools.ipgp.fr/>
.SH "REFERENCES"
.IX Header "REFERENCES"
Wieczorek, M. A. and R. J. Phillips, Potential anomalies on a sphere: applications to the thickness of the lunar crust, \fIJ. Geophys. Res.\fR, 103, 1715\-1724, 1998.
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2012 by Mark Wieczorek <wieczor@ipgp.fr>.
.PP
This is free software; you can distribute and modify it under the terms of the revised \s-1BSD\s0 license.


src/Hilm.f95
subroutine Hilm(cilm, ba, grid, lmax, nmax, mass, r0, rho, gridtype, w, plx, zero, filter_type, filter_deg, lmax_calc)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	This routine will compute the next estimate of Moho coefficients 
!	given an initial estimate of the moho in a gridded data file. This
!	is simply Equation 18 in Wieczorek and Phillips (1998). Note that the
!	degree-0 topography must be included in the gridded relief. Note that 
!	the array plx is optional, and should not be precomputed when memory 
!	is an issue (i.e., lmax>360).
!
!	Calling Parameters:
!		IN
!			ba:		Bouguer Anomaly spherical harmonic coefficients.
!			grid:	Initial estimate of Moho refief, gridded according
!					to a call to Makegrid.
!			lmax:		Maxmimum spherical harmonic degree to compute.
!			nmax:		Order of potential coefficient expansion.
!			mass:		Mass of planet.	
!			R0:		The radius that the coefficients are referenced
!					to. 		
!			rho:		Density contrast between the mantle and crust.
!			gridtype	1 = Gauss-Legendre quadrature grid corresponding to LMAX.
!					2 = N by N Driscoll and Healy grid corresponding to LMAX.
!					3 = N by 2N Driscoll and Healy grid corresponding to LMAX.
!		OUT
!			cilm:		Estimate of Moho relief spherical harmonic coefficients, 
!					with dimensions (2, lmax+1, lmax+1).
!			
!		OPTIONAL		
!			w:		Gauss-Legendre points used in integrations (for GRIDTYPE=1, 
!					determined from a call to PreCompute).
!			zero:		Array of dimension lmax+1 that contains the latitudinal
!					gridpoints used in the Gauss-Legendre quadrature integration
!					scheme (For GRIDTYPE=1). Only needed if plx is not included.
!			plx:		Input array of Associated Legendre Polnomials computed
!					at the Gauss points (for GRIDTYPE=1, determined from a call to
!					PreCompute). If this is not included, then the optional
!					array zero MUST be inlcuded.
!			filter_deg:	If specified, each interation will be filtered according to
!					equation 18 in Wieczorek and Phillips (1998), where the value
!					of filter_deg corresponds to the spherical harmonic degree where 
!					the filter is 1/2.
!			filter_type:	If filter_deg is specified, this must be as well. 
!					A value of (0) corresponds to none, (1) corresponds to the minimum amplitude filter in 
!					Wieczorek and Phillips (1998), whereas as value of (2) corresponds
!					to a minimum curvature filter.
!			lmax_calc:	Maximum degree to compute spherical harmonic expansions up to.
!
!
!	All units assumed to be SI.
!
!	Dependencies:		NGLQSH, SHExpandGLQ, SHExpandDH, wl, wl_curv
!
!	Written by Mark Wieczorek 2003
!		September 3, 2005. Modifed so that the array plx is now optional.
!		May 19, 2010. Modified to use Dricoll and Healy grids.
!		May 15, 2012. Modifed so that the first spherical harmonic expansion only computes the degree 0 term.
!
!	Copyright (c) 2005-2010, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	use SHTOOLS, only: NGLQSH, SHExpandGLQ, SHExpandDH, Wl, WlCurv

	implicit none
		
	real*8, intent(out) :: 			cilm(:,:,:)
	real*8, intent(in) ::			ba(:,:,:), grid(:,:), mass, r0, rho
	real*8, intent(in), optional ::		plx(:,:), zero(:), w(:)
	integer, intent(in) ::			lmax, nmax, gridtype
	integer, intent(in), optional :: 	filter_type, filter_deg, lmax_calc
	real*8 ::				prod, pi, d, depth, filter(lmax+1)
	real*8, allocatable ::			cilmn(:, :, :), grid2(:,:)
	integer  ::				j, l, n, nlong, nlat, astat(2), lmax_out, lmax_calc2

	if (present(lmax_calc)) then
		if (lmax_calc > lmax .or. lmax_calc < 0) then
			print*, "Error -- Hilm"
			print*, "LMAX_CALC must be less than or equal to LMAX."
			print*, "LMAX = ", lmax
			print*, "LMAX_CALC = ", lmax_calc
			stop
		endif
		lmax_calc2 = lmax_calc
	else
		lmax_calc2 = lmax
	endif

	if (size(cilm(:,1,1)) < 2 .or. size(cilm(1,:,1)) < lmax_calc2+1 .or. size(cilm(1,1,:)) < lmax_calc2+1) then
		print*, "Error --- Hilm"
		print*, "CILM must be dimensioned as (2, LMAX+1, LMAX+1) where LMAX is ", lmax_calc2
		print*, "Input dimension is ", size(cilm(:,1,1)), size(cilm(1,:,1)), size(cilm(1,1,:))
		stop
	elseif (size(ba(:,1,1)) < 2 .or. size(ba(1,:,1)) < lmax_calc2+1 .or. size(ba(1,1,:)) < lmax_calc2+1) then
		print*, "Error --- Hilm"
		print*, "BA must be dimensioned as (2, LMAX+1, LMAX+1) where LMAX is ", lmax_calc2
		print*, "Input dimension is ", size(ba(:,1,1)), size(ba(1,:,1)), size(ba(1,1,:))
		stop
	endif
	
	if (gridtype == 1) then
		nlat = NGLQSH(lmax) ! lmax+1
		nlong = 2*lmax+1
	elseif (gridtype == 2) then 
		nlat = 2*lmax+2
		nlong = nlat
	elseif (gridtype == 3) then
		nlat = 2*lmax+2
		nlong = 2*nlat
	else
		print*, "Error --- Hilm"
		print*, "GRIDTYPE must be 1 (Gauss-Legendre Quadrature), 2 (Driscoll-Healy NxN) or 3 (Driscoll-Healy Nx2N)"
		stop
	endif
	
	
	if(size(grid(1,:)) < nlong .or. size(grid(:,1)) <  nlat) then
		print*, "Error --- Hilm"
		print*, "grid must be dimensioned as (NLAT, NLONG) where"
		print*, "NLAT = ", nlat
		print*, "NLONG = ", nlong
		print*, "Input dimension is ", size(grid(1,:)), size(grid(:,1))
		stop
	endif
	
	if (present(w)) then
		if (gridtype /= 1) then
			print*, "Error --- Hilm"
			print*, "W can only be present when GRIDTYPE is 1 (Gauss-Legendre Quadrature)"
			print*, "GRIDTYPE = ", gridtype
			stop
		elseif(size(w) < lmax+1) then
			print*, "Error --- Hilm"
			print*, "W must be dimensioned as (LMAX+1) where LMAX is ", lmax
			print*, "Input dimension is ", size(w)
			stop
		endif
	endif

	if (present(zero)) then
		if (gridtype /= 1) then
			print*, "Error --- Hilm"
			print*, "ZERO can only be present when GRIDTYPE is 1 (Gauss-Legendre Quadrature)"
			print*, "GRIDTYPE = ", gridtype
			stop
		elseif (size(zero) < lmax + 1) then
			print*, "Error --- Hilm"
			print*, "ZERO must be dimensioned as (LMAX+1) where LMAX is ", lmax
			print*, "Input dimension is ", size(zero)
			stop
		endif
	endif
	
	if (present(plx)) then
		if (gridtype /= 1) then
			print*, "Error --- Hilm"
			print*, "PLX can only be present when GRIDTYPE is 1 (Gauss-Legendre Quadrature)"
			print*, "GRIDTYPE = ", gridtype
			stop
		elseif (size(plx(:,1)) < lmax+1 .or. size(plx(1,:)) < (lmax+1)*(lmax+2)/2) then
			print*, "Error --- Hilm"
			print*, "PLX must be dimensioned as (LMAX+1, (LMAX+1)*(LMAX+2)/2) where LMAX is ", lmax
			print*, "Input dimension is ", size(plx(:,1)), size(plx(1,:))
			stop
		endif
	endif

	if (present(w)) then
		if (present(plx) .and. present(zero)) then
			print*, "Error --- Hilm"
			print*, "If the optional parameter W is present, then only one of PLX and ZERO must also be present."
			stop
		endif
		if (.not.present(plx) .and. .not.present(zero)) then
			print*, "Error --- Hilm"
			print*, "If the optional parameter W is present, then one of PLX or ZERO must also be present."
			stop
		endif
	endif
	
	if(present(filter_type)) then
		if (filter_type /=0 .and. filter_type /=1 .and. filter_type /=2) then
			print*, "Error --- Hilm"
			print*, "FILTER_TYPE must be either 0 (none), 1 (minimum amplitude), or 2 (minimum curvature)"
			print*, "Input value is ", filter_type
			stop
		endif
		if (filter_type == 1 .or. filter_type == 2) then
			if (.not.present(filter_deg)) then
				print*, "Error --- Hilm"
				print*, "If FILTER_TYPE is present and equal to 1 or 2, so mush FILTER_DEG."
				stop
			endif
		endif
	endif
	
	allocate(cilmn(2, lmax_calc2+1, lmax_calc2+1), stat = astat(1))
	allocate(grid2(nlat, nlong), stat = astat(2))
	if (astat(1) /= 0 .or. astat(2) /= 0) then
		print*, "Error --- Hilm"
		print*, "Problem allocating arrays CILMN and GRID2", astat(1), astat(2)
		stop
	endif

	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Do the expansions
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	cilm = 0.0d0
	cilmn = 0.0d0
	pi = acos(-1.0d0)
	
	grid2(1:nlat,1:nlong) = grid(1:nlat,1:nlong)
	
	if (gridtype == 1) then
		if (present(plx)) then
			call SHExpandGLQ(cilmn, lmax, grid2(1:nlat,1:nlong), w(1:lmax+1), &
				plx = plx(1:lmax+1, 1:(lmax+1)*(lmax+2)/2), norm = 1, csphase = 1, lmax_calc = 0)
		else 
			call SHExpandGLQ(cilmn, lmax, grid2(1:nlat,1:nlong), w(1:lmax+1), &
				zero = zero(1:lmax+1), norm = 1, csphase = 1, lmax_calc = 0)
		endif
	elseif (gridtype == 2) then
		call SHExpandDH(grid2(1:nlat,1:nlong), nlat, cilmn, lmax_out, norm = 1, &
			sampling = 1, csphase = 1, lmax_calc = 0)
	elseif (gridtype == 3) then
		call SHExpandDH(grid2(1:nlat,1:nlong), nlat, cilmn, lmax_out, norm = 1, &
			sampling = 2, csphase = 1, lmax_calc = 0)
	endif

	d = cilmn(1,1,1)
	depth = r0-d
	print*, "Average depth of Moho (km) = ", depth/1.d3
	
	cilm(1,1,1) = d
		
	filter = 1.0d0
	if (present(filter_type) .and. present(filter_deg)) then
		do l=1, lmax_calc2
			if (filter_type == 1) then
				filter(l+1) = Wl(l, filter_deg, r0, d)
			elseif(filter_type == 2) then
				filter(l+1) = WlCurv(l, filter_deg, r0, d)
			endif
		enddo
	endif
	
	do l=1, lmax_calc2
		cilm(1:2,l+1,1:l+1) = filter(l+1)*ba(1:2,l+1,1:l+1) * mass * dble(2*l+1) * ( (r0/d)**l) / &
			(4.0d0*pi*rho*d**2)
	enddo
		
	do n=2, nmax
		
		grid2(1:nlat,1:nlong) = ((grid(1:nlat,1:nlong) - d)/d)**n

		if (gridtype == 1) then
			if (present(plx)) then
				call SHExpandGLQ(cilmn, lmax, grid2(1:nlat,1:nlong), w(1:lmax+1), &
					plx = plx(1:lmax+1, 1:(lmax+1)*(lmax+2)/2), norm = 1, csphase = 1, lmax_calc = lmax_calc2)
			else
				call SHExpandGLQ(cilmn, lmax, grid2(1:nlat,1:nlong), w(1:lmax+1), &
					zero = zero(1:lmax+1), norm = 1, csphase = 1, lmax_calc = lmax_calc2)
			endif
		elseif (gridtype == 2) then
			call SHExpandDH(grid2(1:nlat,1:nlong), nlat, cilmn, lmax_out, norm = 1, &
				sampling = 1, csphase = 1, lmax_calc = lmax_calc2)
		elseif (gridtype == 3) then
			call SHExpandDH(grid2(1:nlat,1:nlong), nlat, cilmn, lmax_out, norm = 1, &
				sampling = 2, csphase = 1, lmax_calc = lmax_calc2)
		endif
		
		do l = 1, lmax_calc2
			prod = 1.0d0
			do j=1, n
				prod = prod * dble(l+4-j)
			enddo
			prod = d * prod/( dble(l+3) * dble(fact(n)) )
			
			cilm(1:2,l+1,1:l+1) = cilm(1:2,l+1,1:l+1) - filter(l+1)*cilmn(1:2,l+1,1:l+1)*prod
		enddo
	enddo
	
	deallocate(cilmn)
	deallocate(grid2)	
	
	contains
	
		function fact(i)
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		!
		!	This function computes the factorial of an integer
		!
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
			implicit none
			integer ::	i, j, fact
	
			if (i == 0) then
				fact = 1	
			elseif (i .lt. 0) then
				print*, "Argument to FACT must be positive"
				stop
			else
				fact = 1
				do j = 1, i
					fact = fact * j
				enddo
			endif
			
		end function fact
	
end subroutine Hilm

